Token rendering, text canonicalization, minimal test suite.

fisx opened the pull request:
As promised long ago in #2.  Any feedback welcome.

I've introduced token rendering and canonicalization because I'm using it for the tests and in my production code (not released).  If you don't want it to clutter the api, I can move it to tests.

I'm also hoping to release some code that forms trees from token streams (or errors).  Should I try another PR for that or should I go for a separate library?  (I'm assuming the latter, but I'm always happy to dump my code on other maintainers.  ;-)

fisx committed:
Token rendering, text canonicalization, minimal test suite.

bgamari commented on Text/HTML/Parser.hs:
```diff
@@ -14,6 +15,12 @@ module Text.HTML.Parser
     , Token(..)
     , TagName, AttrName, AttrValue
     , Attr(..)
+      -- * Rendering, text canonicalization
```
I wonder whether we should split things up a bit more. This would involve moving the types (and perhaps canonicalization) to a `.Types` module, moving the parser to a `.Parser` module and the rendering to a `.Render`module.

However, what you have done here is fine for now.

bgamari reviewed (approved):
Thanks for this; it looks like a great start!

Another mode of testing that I've been thinking of adding recently is a set of golden tests with some "typical" pages and their corresponding token streams. I've had good luck with `tasty-silver` for this purpose in the past. Perhaps I'll give this a shot soon.

As far as additional parsing support is concerned, I wouldn't be opposed taking patches for this as long as it didn't go much farther than an element tree. I was considering implementing this myself but found it was possible to do the extraction that I needed from the token stream.

fisx committed:
Merge branch 'master' into tests-and-things

Conflicts:
	html-parse.cabal

fisx commented:
Thanks for the nice feedback!  I'll see if I can find the time soon to split up the module and add one for tree parsing (see #5).

I'll keep working the travis errors a bit more now.

fisx committed:
Relax base constraint.

fisx committed:
Add hspec-discover dep.

It's a bit of a particular dependency because we need an executable,
not a library, but I think this might still work.

fisx commented:
Oh, I just saw you do some sophisticated travis hacking to avoid redundant test runs.  Do you have a hint what I should do?

fisx committed:
Always install hspec-discover (travis hack).

fisx commented:
Now the test suite is building ok and runs, but does not run any tests.  Odd.

I'm out of time for today.  Any hints appreciated!

bgamari commented:
Hmm, I wonder what is going on here. I'll have a look tomorrow.

fisx commented:
I think I've fixed it in #6.
